File: android/src/main/java/com/lightningkite/kwift/observables/binding/CompoundButton.binding.actual.kt
package com.lightningkite.kwift.observables.binding

import android.widget.CompoundButton
import com.lightningkite.kwift.observables.*


fun <T> CompoundButton.bindSelect(value: T, observable: MutableObservableProperty<T>) {
    observable.addAndRunWeak(this) { self, it ->
        val shouldBeChecked = it == value
        if (self.isChecked != shouldBeChecked) {
            self.isChecked = shouldBeChecked
        }
    }
    setOnCheckedChangeListener { buttonView, isChecked ->
        if (isChecked && observable.value != value) {
            observable.value = value
        } else if (!isChecked && observable.value == value) {
            this.isChecked = true
        }
    }
}

fun <T> CompoundButton.bindSelectNullable(value: T, observable: MutableObservableProperty<T?>) {
    observable.addAndRunWeak(this) { self, it ->
        val shouldBeChecked = it == value
        if (self.isChecked != shouldBeChecked) {
            self.isChecked = shouldBeChecked
        }
    }
    setOnCheckedChangeListener { buttonView, isChecked ->
        if (isChecked && observable.value != value) {
            observable.value = value
        } else if (!isChecked && observable.value == value) {
            observable.value = null
        }
    }
}

fun <T> CompoundButton.bindSelectInvert(value: T, observable: MutableObservableProperty<T?>) {
    var suppress = false
    observable.addAndRunWeak(this) { self, it ->
        if (!suppress) {
            suppress = true
            val shouldBeChecked = it == value || it == null
            if (self.isChecked != shouldBeChecked) {
                self.isChecked = shouldBeChecked
            }
            suppress = false
        }
    }
    setOnCheckedChangeListener { buttonView, isChecked ->
        if (!suppress) {
            suppress = true
            if (!isChecked && observable.value == value) {
                observable.value = null
                buttonView.isChecked = true
            } else if (observable.value != value) {
                observable.value = value
                buttonView.isChecked = true
            }
            suppress = false
        }
    }
}

fun CompoundButton.bind(observable: MutableObservableProperty<Boolean>) {
    observable.addAndRunWeak(this) { self, it ->
        if (it != this.isChecked) {
            self.isChecked = it
        }
    }
    setOnCheckedChangeListener { buttonView, isChecked ->
        if (observable.value != isChecked) {
            observable.value = isChecked
        }
    }
}

File: ios/Kwift/com/lightningkite/kwift/observables/binding/CompoundButton.binding.actual.swift
//Stub file made with Kwift 2 (by Lightning Kite)
import Foundation


//--- CompoundButton.bindSelect(T, MutableObservableProperty<T>)
public extension CompoundButton {
    func bindSelect<T: Equatable>(_ myValue: T, _ selected: MutableObservableProperty<T>) -> Void {
        selected.addAndRunWeak(referenceA: self) { (this, value) in
            let shouldBeOn = value == myValue
            if this.isOn != shouldBeOn {
                this.isOn = shouldBeOn
            }
        }
        self.onCheckChanged = { [weak self] value in
            if value && selected.value != myValue {
                selected.value = myValue
            } else if !value && selected.value == myValue  {
                self?.isOn = true
            }
        }
    }
    func bindSelect<T: Equatable>(value: T, observable: MutableObservableProperty<T>) -> Void {
        return bindSelect(value, observable)
    }
}

//--- CompoundButton.bindSelectNullable(T, MutableObservableProperty<T?>)
public extension CompoundButton {
    func bindSelectNullable<T: Equatable>(_ myValue: T, _ selected: MutableObservableProperty<T?>) -> Void {
        selected.addAndRunWeak(referenceA: self) { (this, value) in
            let shouldBeOn = value == myValue
            if this.isOn != shouldBeOn {
                this.isOn = shouldBeOn
            }
        }
        self.onCheckChanged = { [weak self] value in
            if value && selected.value != myValue {
                selected.value = myValue
            } else if !value && selected.value == myValue  {
                selected.value = nil
            }
        }
    }
    func bindSelectNullable<T: Equatable>(value: T, observable: MutableObservableProperty<T?>) -> Void {
        return bindSelectNullable(value, observable)
    }
}

//--- CompoundButton.bindSelectInvert(T, MutableObservableProperty<T?>)
public extension CompoundButton {
    func bindSelectInvert<T: Equatable>(_ myValue: T, _ selected: MutableObservableProperty<T?>) -> Void {
        var suppress:Bool = false
        selected.addAndRunWeak(referenceA: self) { (this, value) in
            if !suppress{
                suppress = true
                let shouldBeOn = value == myValue || value == nil
                if this.isOn != shouldBeOn {
                    this.isOn = shouldBeOn
                }
                suppress = false
            }
        }
        self.onCheckChanged = { [weak self] value in
            if !suppress{
                suppress = true
                if !value && selected.value == myValue {
                    selected.value = nil
                    self?.isOn = true
                } else if selected.value != myValue  {
                    selected.value = myValue
                    self?.isOn = true
                }
                suppress = false
            }
        }
    }
    func bindSelectInvert<T: Equatable>(value: T, observable: MutableObservableProperty<T?>) -> Void {
        return bindSelectInvert(value, observable)
    }
}

//--- CompoundButton.bind(MutableObservableProperty<Boolean>)
public extension CompoundButton {
    func bind(_ observable: MutableObservableProperty<Bool>) -> Void {
        observable.addAndRunWeak(referenceA: self) { (this, value) in
            if this.isOn != value {
                this.isOn = value
            }
        }
        self.onCheckChanged = { [weak self] value in
            if observable.value != self?.isOn {
                observable.value = self?.isOn ?? false
            }
        }
    }
    func bind(observable: MutableObservableProperty<Bool>) -> Void {
        return bind(observable)
    }
}






